suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts

```diff
@@ -0,0 +1,290 @@
+            coin,
+            descriptor,
+            pageSize: TRANSACTION_ADD_ACCOUNT_PAGE_SIZE,
+        });
+        this.watchForNewTransactions({ coin, descriptor });
+    }
+
+    async accountExists({ coin, descriptor }: AccountUniqueParams) {
+        await this.initPromise;
+
+        return this.storage.accountExist({ coin, descriptor });
+    }
+
+    async removeAccount({ coin, descriptor }: AccountUniqueParams) {
+        await this.initPromise;
+        await this.storage.removeAccount({ coin, descriptor });
+        await this.stopWatchingForNewTransactions({ coin, descriptor });
+    }
+
+    @WaitUntilFetchIsFinished()
+    public async getTransactions({ coin, descriptor }: AccountUniqueParams) {
+        debugLog(`getTransactions for ${coin} - ${descriptor}`);
+        await this.initPromise;
+
+        return this.storage.getTransactions({ coin, descriptor });
+    }
+
+    @EnsureSingleRunningInstance()
+    private async fetchAndSaveAllMissingAccountTransactions({
+        coin,
+        descriptor,
+        pageSize,
+    }: AccountUniqueParams & { pageSize: number }) {
+        await this.initPromise;
+        debugLog(`Fetching transactions for ${coin} - ${descriptor}`, { pageSize });
+        if (!(await this.accountExists({ coin, descriptor }))) {
+            throw new Error(
+                'Account not registered. Please register it first using TransactionCacheEngine.addAccount',
+            );
+        }
+
+        const allTransactions: AccountTransaction[] = [];
+
+        let page = 0;
+        // marker is used instead of page for ripple (cursor based pagination)
+        let marker: AccountInfo['marker'] | undefined;
+        const networkType = getNetworkType(coin);
+        // eslint-disable-next-line no-constant-condition
+        while (true) {
+            page += 1;
+
+            const result = await TrezorConnect.getAccountInfo({
+                coin,
+                descriptor,
+                details: 'txs',
+                page,
+                pageSize,
+                ...(marker ? { marker } : {}), // set marker only if it is not undefined (ripple), otherwise it fails on marker validation
+            });
+            if (!result.success) {
+                throw new Error(result.payload.error);
+            }
+
+            marker = result.payload.marker;
+
+            const pageTransactions = result.payload.history.transactions ?? [];
+
+            allTransactions.push(...pageTransactions);
+
+            const txids = pageTransactions.map(tx => tx.txid);
+            const numberOfExistingTransactions = await this.storage.getNumberOfExistingTransactions(
+                {
+                    coin,
+                    descriptor,
+                    txids,
+                },
+            );
+
+            // We have fetched all transactions that are not already saved in the storage.
+            if (numberOfExistingTransactions === txids.length) {
```

Comment:
I don't get this condtion. How does the storage know, that it has all the missing transactions?

Reply:
Yea, this one it's pretty tricky if I see it now. I added better comment.

Basically you call `this.storage.getNumberOfExistingTransactions` on latest page of transaction fetched, if all exists you know you don't need to go to next page because you have already fetched all missing transactions.